src/nvim/tui/tui.c

```diff
@@ -0,0 +1,718 @@
+  data->fg = data->bg = -1;
+  data->can_use_terminal_scroll = true;
+  data->bufpos = 0;
+  data->option_cache = pmap_new(cstr_t)();
+
+  // write output to stderr if stdout is not a tty
+  data->out_fd = isatty(1) ? 1 : (isatty(2) ? 2 : 1);
+  kv_init(data->invalid_regions);
+  // setup term input
+  data->input = term_input_new();
+  // setup unibilium
+  data->ut = unibi_from_env();
+  if (!data->ut) {
+    // For some reason could not read terminfo file, use a dummy entry that
+    // will be populated with common values by fix_terminfo below
+    data->ut = unibi_dummy();
+  }
+  fix_terminfo(data);
+  // Enter alternate screen and clear
+  unibi_out(ui, unibi_enter_ca_mode, NULL);
+  unibi_out(ui, unibi_clear_screen, NULL);
+
+  // setup output handle in a separate event loop(we wanna do synchronous
+  // write to the tty)
+  data->write_loop = xmalloc(sizeof(uv_loop_t));
+  uv_loop_init(data->write_loop);
+  uv_tty_init(data->write_loop, &data->output_handle, data->out_fd, 0);
+
+  // Obtain screen dimensions
+  update_size(ui);
+
+  // listen for SIGWINCH
+  uv_signal_init(uv_default_loop(), &data->winch_handle);
+  uv_signal_start(&data->winch_handle, sigwinch_cb, SIGWINCH);
+  data->winch_handle.data = ui;
+
+  ui->stop = tui_stop;
+  ui->rgb = false;
+  ui->data = data;
+  ui->resize = tui_resize;
+  ui->clear = tui_clear;
+  ui->eol_clear = tui_eol_clear;
+  ui->cursor_goto = tui_cursor_goto;
+  ui->cursor_on = tui_cursor_on;
+  ui->cursor_off = tui_cursor_off;
+  ui->mouse_on = tui_mouse_on;
+  ui->mouse_off = tui_mouse_off;
+  ui->insert_mode = tui_insert_mode;
+  ui->normal_mode = tui_normal_mode;
+  ui->set_scroll_region = tui_set_scroll_region;
+  ui->scroll = tui_scroll;
+  ui->highlight_set = tui_highlight_set;
+  ui->put = tui_put;
+  ui->bell = tui_bell;
+  ui->visual_bell = tui_visual_bell;
+  ui->update_fg = tui_update_fg;
+  ui->update_bg = tui_update_bg;
+  ui->flush = tui_flush;
+  ui->suspend = tui_suspend;
+  ui->set_title = tui_set_title;
+  ui->set_icon = tui_set_icon;
+  // Attach
+  ui_attach(ui);
+}
+
+static void tui_stop(UI *ui)
+{
+  TUIData *data = ui->data;
+  // Destroy common stuff
+  kv_destroy(data->invalid_regions);
+  uv_signal_stop(&data->winch_handle);
+  uv_close((uv_handle_t *)&data->winch_handle, NULL);
+  // Destroy input stuff
+  term_input_destroy(data->input);
+  // Destroy output stuff
+  tui_normal_mode(ui);
+  tui_mouse_off(ui);
+  unibi_out(ui, unibi_exit_attribute_mode, NULL);
+  unibi_out(ui, unibi_cursor_normal, NULL);
+  unibi_out(ui, unibi_exit_ca_mode, NULL);
+  flush_buf(ui);
+  uv_close((uv_handle_t *)&data->output_handle, NULL);
+  uv_run(data->write_loop, UV_RUN_DEFAULT);
+  if (uv_loop_close(data->write_loop)) {
+    abort();
+  }
+  free(data->write_loop);
+  unibi_destroy(data->ut);
+  char *opt_value;
+  map_foreach_value(data->option_cache, opt_value, {
+    free(opt_value);
+  });
+  pmap_free(cstr_t)(data->option_cache);
+  ui_detach(ui);
+}
```

Comment:
`tui_stop()` needs to free some more pointers. Memory leaks can add up when `tui_suspend()` is called a few times.
```diff
diff --git a/src/nvim/tui/tui.c b/src/nvim/tui/tui.c
index 92e35e5..527954e 100644
--- a/src/nvim/tui/tui.c
+++ b/src/nvim/tui/tui.c
@@ -175,6 +175,16 @@ static void tui_stop(UI *ui)
     free(opt_value);
   });
   pmap_free(cstr_t)(data->option_cache);
+
+  if (data->screen) {
+    for (int i = 0; i < data->old_height; i++) {
+      free(data->screen[i]);
+    }
+    free(data->screen);
+  }
+  free(data);
+  free(ui);
+
   ui_detach(ui);
 }
 
```